### ISSUE TITLE
MouseOver event for entire stacked bar, not each small rectangle in the StackedBarChart

### ISSUE THREAD

**Issue by corinnebrady** [recharts/recharts#759]
I need to be able to trigger a mouse event from the entire vertical bar of a Stacked Bar Chart.  

Currently, if you add `onMouseEnter` on on `<Bar ...` it triggers when you mouse over the corresponding colour on each bad (which makes sense), but isn't what I need.

When you add a `<Tooltip ...` it is able to trigger when on a single bar, which is what I need, but i don't need the Recharts tooltip.  I need to be able to call a function that triggers state within my React component.  

### My code:
```
        <BarChart
          width={205}
          height={275}
          barCategoryGap="25%"
          data={graphData}
          margin={{ top: 20, right: 15, left: 0, bottom: 5 }}
        >
          <XAxis
            dataKey="name"
            tickLine={false}
            tick={<XAxisTick />}
            stroke={color.grey400}
          />
          <Bar
            dataKey="project"
            stackId="a"
            fill={color.blueMedium}
            isAnimationActive={false}
            label={<BarLabel currency={currencySymbol} />}
            onMouseEnter={this.handleShowDivA}
            onMouseLeave={this.handleHideDivA}
          />
          <Bar
            dataKey="reccuring"
            stackId="a"
            fill={color.blueDark}
            isAnimationActive={false}
            onMouseEnter={this.handleShowDivB}
            onMouseLeave={this.handleHideDivB}
          />
        </BarChart>
```

### What currently happens when using `onMouseEnter` etc:
![current mouse events](https://user-images.githubusercontent.com/5469814/27778530-e846c1d6-5ffe-11e7-8cca-db59636d8c35.jpg)

### What I need to happen with a `onMouseEnter`:
![desired mouse events](https://user-images.githubusercontent.com/5469814/27778548-0baa7ac8-5fff-11e7-8acd-72ac0768fdfc.jpg)

Does anyone know how to do this? Is it even possible?

**Comment by xile611** [comment]
@corinnebrady  I don't know why your problem can't be solved by triggering the same event when you mouse enter Bar  1 and Bar 2.

[demo](http://jsfiddle.net/fm1qLLsz/)

**Comment by corinnebrady** [comment]
@xile611 I've updated the JS Fiddle to should you want I mean.  Hover over a single bar and switch between the colours.  You'll see in console that the console.log changes.  Now try hovering over each green portion of multiple bars, then hover over the purple portion of multiple bars.  

Essentially, every time the green portion is hovered over on every single bar it calls one console.log, and every time the purple portion is hovered over on every single bar it calls the other console log.  

What I need is a single mouse even for each vertical bar no matter the colour.  I need a mouse even for `Page A` bar, `Page B` bar, etc.  

Any ideas?

**Comment by neu5** [comment]
@corinnebrady did you resolved this?

**Comment by corinnebrady** [comment]
@neu5 No I never did.  Honestly, I ended up building this chart myself from scratch with D3 instead.

**Comment by neu5** [comment]
@corinnebrady I try to understand what problem you had, is something like that a solution?
https://codesandbox.io/s/n15x65pw4j

**Comment by yuriploc** [comment]
@corinnebrady I'm facing the same problem, have been trying for a whole day to change the stacked bar.

@neu and @xile611, The codepen (http://jsfiddle.net/fm1qLLsz/) is broken. Anyway, how would you get only the active bar on mouseover and change its style?

**Comment by neu5** [comment]
@yuriploc did you check out my codesandbox.io?
https://codesandbox.io/s/n15x65pw4j

**Comment by yuriploc** [comment]
@neu5 Yes, I checked your code, it's not useful for the problem I need to solve. The object is read-only and has no style attrs.

The solution I had was to use the `Cell` and vanilla JS. The problem is that the stacked bar renders the bottom bars, then it renders the top bars, and we can't access _a_ stack as an element.

**Comment by neu5** [comment]
@yuriploc yup, unfortunately that's the way it's implemented.

**Comment by erosenberg** [comment]
Is there still no way to access a stack as a single element or component? Is there a way to get all of the data one would need from the props by each stack? That would be incredibly helpful if BarChart's props for example contained an array of stacks with corresponding data/sums/colors/values, etc.

**Comment by ahmedramadan65** [comment]
im facing the same problem .. did anyone get solution for this ?

**Comment by yuriploc** [comment]
@ahmedramadan65 

> The solution I had was to use the Cell and vanilla JS. The problem is that the stacked bar renders the bottom bars, then it renders the top bars, and we can't access a stack as an element.

**Comment by Haraldson** [comment]
The solution looks like this:

```javascript
import React, { useState } from 'react'
import { BarChart, Bar, Rectangle } from 'recharts'

const BarChartContainer = ({ data, labels, colors, ...whatnot }) => {
    const [activeBarLabel, setActiveBarLabel] = useState(null)

    return (
        <BarChart data={data}>
            {labels.map((label, key, index) => (
                <Bar
                    dataKey={label}
                    stackId="x"
                    fill={colors[index]}
                    shape={props => {
                        const active = props.label === activeBarLabel

                        props = {
                            ...props,
                            width: active ? 50 : 40
                        }

                        return <Rectangle {...props} />
                    }}
                    onMouseEnter={({ label }) => setActiveBarLabel(label)}
                    onMouseLeave={() => setActiveBarLabel(null)}
                    key={label} />
            ))}
        </BarChart>
    )
}

export default BarChartContainer
```

**Comment by AbhaysinghBhosale** [comment]
onMouseEnter={({ label }) => setActiveBarLabel(label)}
onMouseLeave={() => setActiveBarLabel(null)}

This events are not working after updating to version 2. Anyone have found any solution for this.

**Comment by antony-chiramel-infy** [comment]
I have found a solution to this problem.
onMouseEnter={(event) => handleLineHover(event, label)}

Inside your handleLineHover

const handleLineHover = (event, label) => {
       setActiveBarLabel(label)
}

That's it!!

**Comment by yiyb0603** [comment]
> The solution looks like this:
> 
> ```js
> import React, { useState } from 'react'
> import { BarChart, Bar, Rectangle } from 'recharts'
> 
> const BarChartContainer = ({ data, labels, colors, ...whatnot }) => {
>     const [activeBarLabel, setActiveBarLabel] = useState(null)
> 
>     return (
>         <BarChart data={data}>
>             {labels.map((label, key, index) => (
>                 <Bar
>                     dataKey={label}
>                     stackId="x"
>                     fill={colors[index]}
>                     shape={props => {
>                         const active = props.label === activeBarLabel
> 
>                         props = {
>                             ...props,
>                             width: active ? 50 : 40
>                         }
> 
>                         return <Rectangle {...props} />
>                     }}
>                     onMouseEnter={({ label }) => setActiveBarLabel(label)}
>                     onMouseLeave={() => setActiveBarLabel(null)}
>                     key={label} />
>             ))}
>         </BarChart>
>     )
> }
> 
> export default BarChartContainer
> ```

this code works for me thanks :)

**Comment by nikolasrieble** [comment]
Given that a workaround exists - thank you for sharing @Haraldson - I think the only thing missing here is documentation for now.
